src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -532,6 +544,142 @@ public static int GetCurrentProcessorId()
             return currentProcessorIdCache >> ProcessorIdCacheShift;
         }
 
+        // do a fast check and record in a readonly static so that it could become a JIT constant
+        internal static readonly bool IsCoreIdReallyFast = SimpleCoreIdSpeedCheck();
+
+        // If GetCurrentProcessorNumber takes any nontrivial time (compared to TLS access), return false.
+        // Check more than once - to make sure it was not because TLS was delayed by GC or a context switch.
```

Comment:
If there is a GC or context switch, there is a pretty good chance to get a false hit on the `(t3 - t2) * 2 < t1` check below. No?

Reply:
context switch or GC is a possibility. If we are unlucky, we will calibrate, which is more reliable.
 
In practice it does not seem to happen often. This loop is fast. In positive cases the API is cheap. In negative cases we fail in 1 or 2 iterations. The time cap is just for unusual "slow positive case" scenario.

---
In reply to: [352963343](https://github.com/dotnet/runtime/pull/467#discussion_r352963343) [](ancestors = 352963343)

Reply:
I tried running a test app with printf that reports calibration results. Even when machine also runs GC stress results from the app are fairly consistent. Stressing just made everything slower but did not appear to have much effect on likelihood of having false negatives or on results of calibrations.

This loop is limited to 100 usec, but typically takes less and that is intended to be short relative to a typical OS quantum, which is 10-100 msec on modern OSes.

---
In reply to: [352986670](https://github.com/dotnet/runtime/pull/467#discussion_r352986670) [](ancestors = 352986670,352963343)

Reply:
So what is the likelihood of having false negatives from SimpleCoreIdSpeedCheck in the worst case?

I think it is important to have predictable performance. I do not think it is ok for one context switch in a bad spot to affecting performance characteristics for the duration of the process. Getting 5 or 10 consecutive context in bad spots - I can live with that affecting performance characteristics.

Reply:
Now that we have "fast" criteria at 3x, we get `SimpleCoreIdSpeedCheck` working very reliably for positive cases. 

Also, even if false negative happens, it is not a big concern - we will calibrate to some small refresh rate in 0 - 2 range, which is not too bad. 
Such caching is redundant, but will not have a huge impact and will not happen often. 
Two-level checking makes the behavior very robust here.

The opposite scenario could cause more trouble - that is if a very slow `GetProcessorNumber` is accidentally detected as fast. Second level calibration will not help here. 
However for this to happen we must be on a machine with very slow `GetProcessorNumber` in a first place, and `ThreadStatic` access must be interrupted and badly delayed 10 times in a row, while also not causing whole loop to exceed 1 microsecond (since that also triggers calibration path). This is highly improbable scenario.

Reply:
Could you please share the test that you are using to measure `SimpleCoreIdSpeedCheck` reliability, and the results that you are getting?

I have build the following test to measure the reliability under harsh conditions (this is to be compiled directly against System.Private.Corelib - the only change that I have done was to flip SimpleProcessorNumberSpeedCheck to public):

```csharp
using System;
using System.Threading;

class Test
{
    static void Main()
    {
        for (int i = 0; i < 2 * Environment.ProcessorCount; i++)
            new Thread(() => { for (;;) GC.KeepAlive(new object()); }).Start();

        int fast = 0; int slow = 0;
        for (int i = 0; i < 1000; i++)
        {
            if (ProcessorIdCache.SimpleProcessorNumberSpeedCheck()) fast++; else slow++;
        }
        Internal.Console.WriteLine($"{fast} {slow}");
    }
}
```

I get results like:

```
C:\runtime\artifacts\bin\coreclr\Windows_NT.x64.Release>corerun test.exe
703 297
```

It means that this method gives 1 out of 3 times under harsh conditions. That does not look very good. If I replace the early out with a counter (https://github.com/jkotas/runtime/commit/38ba7a415af2a6ad9c55687d1d2cef588310a9f9), I get results that look much better:

```
C:\runtime\artifacts\bin\coreclr\Windows_NT.x64.Release>corerun test.exe
1000 0
```

Reply:
Also, I am curious about the tests and results that prove the two level calibration is really required. It seems we collect enough data in the SimpleProcessorNumberSpeedCheck to calibrate the slow case too.

Reply:
I started with assumptions that false positives are the real hazard. If we have something that is 1000x slower than `ThreadStatic` and accidentally claim that it is "fast" we are in trouble. The opposite is not as bad - if we occasionally have a cache with refresh == 3, that is much less of a problem.
We do not really know how bad CoreID API could perform on everything that we run. Current refresh rate, in a way, supported my concerns.

The other source of uncertainty is timers. Documentation is not very inspiring, Before Vista it was possible to see timers not synchronized between cores for example. Not sure what guarantees we have on other platforms.
The actual frequency and access time are big unknowns. We seem to always report 1ns/tick on Linux. That can't be the actual precision. My older machine reports 284ns/tick on Windows. I suppose it can be easily 10x worse on some other hardware.

We do not want to spend too much time in the static initializer, but we do not want to have false positives. That is why there are two tiers of checks - first is conservative and intended to be a quick detection of decent/recent hardware. When in doubt, we do more expensive, but more reliable measurements. If even that fails we go with default settings.

I wanted the whole deal to be optimized for good hardware, but be resilient to pretty bad cases, if possible.

Reply:
For measurements I build a "printf-instrumented" version of this like in https://github.com/VSadov/runtime/commit/af972102533671a815c4a782a8dbbe0f5d709db0

Then I have an app that uses the API a lot and run it with such runtime. I use the same app that I used originally to measure cached vs. uncached access, but really anything that call the API in a loop will do.

I typically get results like:

```
########## VERY FAST #########
ID: 0.095703125  TLS: 0.040771484375
ID: 0.095703125  TLS: 0.04052734375
ID: 0.095947265625  TLS: 0.040771484375
ID: 0.095703125  TLS: 0.04052734375
ID: 0.095703125  TLS: 0.04052734375
ID: 0.095703125  TLS: 0.04052734375
ID: 0.095703125  TLS: 0.040771484375
ID: 0.095458984375  TLS: 0.04052734375
ID: 0.095703125  TLS: 0.041015625
ID: 0.095703125  TLS: 0.040771484375
Iters: 4096
Refresh: 2

########## VERY FAST #########
ID: 0.11181640625  TLS: 0.0419921875
ID: 0.11181640625  TLS: 0.0419921875
ID: 0.11181640625  TLS: 0.0419921875
ID: 0.11181640625  TLS: 0.0419921875
ID: 0.111328125  TLS: 0.0419921875
ID: 0.11181640625  TLS: 0.04248046875
ID: 0.111328125  TLS: 0.04150390625
ID: 0.11181640625  TLS: 0.0419921875
ID: 0.11181640625  TLS: 0.0419921875
ID: 0.11181640625  TLS: 0.0419921875
Iters: 2048
Refresh: 2
```

or (same machine x86 WOW)
```
ID: 0.7265625, TLS: 0.033203125
ID: 0.7265625, TLS: 0.033203125
ID: 0.724609375, TLS: 0.033203125
ID: 0.7265625, TLS: 0.033203125
ID: 0.7265625, TLS: 0.033203125
ID: 0.7265625, TLS: 0.033203125
ID: 0.724609375, TLS: 0.033203125
ID: 0.724609375, TLS: 0.033203125
ID: 0.7265625, TLS: 0.033203125
ID: 0.724609375, TLS: 0.033203125
Iters: 512
Refresh: 21

ID: 1.21484375, TLS: 0.03515625
ID: 1.21875, TLS: 0.03515625
ID: 0.693359375, TLS: 0.033203125
ID: 0.693359375, TLS: 0.033203125
ID: 0.693359375, TLS: 0.033203125
ID: 0.693359375, TLS: 0.033203125
ID: 0.693359375, TLS: 0.03125
ID: 0.693359375, TLS: 0.03125
ID: 0.693359375, TLS: 0.03125
ID: 0.693359375, TLS: 0.03125
Iters: 512
Refresh: 22
```

under heavy and irregular load the quick check sometimes fails, but calibration still reports refresh rate in the right range.

```
ID: 0.086669921875  TLS: 0.043701171875
ID: 0.490234375  TLS: 0.041015625
ID: 0.2138671875  TLS: 0.0400390625
ID: 0.087890625  TLS: 0.03955078125
ID: 0.087646484375  TLS: 0.03955078125
ID: 0.087646484375  TLS: 0.039306640625
ID: 0.087646484375  TLS: 0.039306640625
ID: 0.087890625  TLS: 0.03955078125
ID: 0.20166015625  TLS: 0.04052734375
ID: 0.088623046875  TLS: 0.040771484375
Iters: 4096
Refresh: 2
```
 As you may see there is a lot of noise even in the second check, but this check is more robust since it has multiple samples needs only one good measurement of each, regardless in which pair.

Reply:
Another source of irregularities that I observed was CPU changing frequency. 
Even in `Ultimate Performance` setting the clock can change quite a bit due to thermal throttling. Second check is ok with that typically, but the first may not - depending when the switch happens.

Reply:
I think we certainly can make the first check more lenient and allow 5 failures (as long as that is not more than 50% of the tries, I would add).

We may need to ask questions though:
- what if `GetProcessorNumber` is ridiculously slow? Can we end up spending too much time in static initializer while re-measuring a hopeless case? 
- what if timer is really bad  (effective frequency is comparable or worse than the cost of `GetProcessorNumber` ) - is that just a lottery of whether clock ticks or not, once or twice, and on what measurement?  Can that lead to labeling real slow `GetProcessorNumber` as "fast" ?